lib/pbench/server/indexer.py

```diff
@@ -1180,42 +1180,59 @@ def _handle_iteration(self, iter_data, iter_name, iter_number, result_json):
         ):
             yield source, PbenchData.make_source_id(source), _parent, _type
 
-    # UID keyword pattern
+    # UID keyword pattern, non-greedy
     _uid_keyword_pat = re.compile(r"%\w*?%")
 
     @staticmethod
-    def expand_template(templ, d, run=None):
-        # match %...% patterns non-greedily.
-        s = templ
-        for m in re.findall(ResultData._uid_keyword_pat, s):
-            # m[1:-1] strips the initial and final % signs from the match.
+    def expand_uid_template(templ, d, run=None):
+        """Given a UID template string using %<keyword>% pattern identifiers,
+        lookup those keywords in the given dictionary and replace those
+        keywords which have values.
+
+        There are two keywords which are treated specially:
+
+          * If the "benchmark_name" is not found in the given dictionary, a
+            second attempt to find a value using "name" will be attempted.
+
+          * If a additional `run` dictionary is provided, it'll be used on a
+            second attempt to find a "controller_host" value by using
+            "controller" in the `run` lookup.
+
+        The supported keyword values are one of: `str`, `int`, or `float`.  If
+        a keyword's value is not a supported type, the replacement is not
+        made.  For floats, 6 decimal places are used in the string conversion.
+
+        Returns a new string with as many keywords replaced as possible.
```

Comment:
The callers never check whether the template is completely expanded, so unexpanded keywords are passed through. Is there any value in detecting that and maybe doing something about it? (although I don't know what we should/could do).

Reply:
If you have a suggestion as to what to do about it, I am all ears.  The UID is a best effort issue and I don't want to fail indexing if they are not all expanded.

Reply:
Is there somewhere suitable where we can log it?

Reply:
We can certainly log it but then we need to spend time constructing a log message with suitable context to be useful.

A query could be used to find all the UID fields still containing the template if someone really cares (you can probably tell where this author is on that spectrum ;).

Reply:
Yeah, it's probably way down there on the importance scale. If it ends up being a problem, maybe we can add such a query to an audit.